core/src/main/scala/kafka/log/Log.scala

```diff
@@ -2254,6 +2256,20 @@ class Log(@volatile var dir: File,
   private[log] def replaceSegments(newSegments: Seq[LogSegment], oldSegments: Seq[LogSegment], isRecoveredSwapFile: Boolean = false): Unit = {
     lock synchronized {
       val sortedNewSegments = newSegments.sortBy(_.baseOffset)
+
+      val (emptyCleanedSegments, nonEmptyCleanedSegments) = sortedNewSegments.partition(_.size == 0)
+
+      if (emptyCleanedSegments.nonEmpty) {
```

Comment:
Currently, we expect that if there is an empty segment, it would be the only segment in newSegments. Perhaps we could throw an IllegalArgumentException if that's not the case.